manpages/pulp_app.1
.TH PULP_APP "1" "July 24" "" "User Commands"
.SH NAME
.B pull_app
\- Bitsream generation, bitstream board download, application compilation and debug interface of PULPissimo project installed with pulp_install.
.SH SYNOPSIS         
.B pulp_app
[-s -d -b [-c APPDIR] [-t APPDIR] -e -h [-u USB|all|i] [-o BOARD] [-r CONNECTOR_NAME]]
.SH DESCRIPTION          
.B pulp_install
is a bash script that should be used after installation of PULPissimo  (possibly using \fBpulp_install\fR) and it is able to create the bitstream of pulpissimo architecture, download it into fpga, compile applications and run it into PULPissimo architecture on fpga using openOCD for degub.  
.PP                    
.SH OPTION
There are two type of option: 
.PP
.B Action 
option that perform a task and are cumulative but executed with a preconfigured order, if all action are selected the script is able to execute them subsequently and terminate, naturally at least one action should be choosen.

.B Config
option instead are used to configure some variable and are optional, this means that these variables have default values.

Now \fB Action \fR option is explained in execution order, this means that if all action are selected the order of execution of action will be this:
.TP 
.B -e|--export-variables
This action export evironment variable previous setted by -b or -c option in .environ.env file.
.TP
.B -s|--bitstream
In order to generate .bit and .bin file this Action enter in pulpissimo directory, it parse the ips_list.yml using the PULP IPApproX IP management tool to generate tcl scripts for all the IPs used in the PULPissimo project using 
.B update-ips
command. The file generated are sourced by Vivado to generate the bitstream for PULPissimo entering in fpga directory and executing 
.B make clean_BOARD
and 
.B make BOARD
where BOARD is the board used setted with -o option or zcu102 as default.
This operation might take a while. If everything goes well your ./pulpissimo/fpga/pulpissimo-BOARD/pulpissimo-BOARD.runs/impl_1 directory should now contain two files:

.B xilinx_pulpissimo.bit
       the bitstream file for JTAG configuration of the FPGA.

.B xilinx_pulpissimo.bin
       the binary configuration file to flash to a non-volatile configuration memory.

This script now support only bitstream download so the architecture should be downloaded every time the board is turned on.
.PP
.B Bitstream common ERROR
 the bitstream generation could not work well due to this kind of problem:
.RS 1.2i
.TP 
.B Vivado version
If bitstream fails, the first cause could be the Vivado version, it should be Vivado 19, not higher. Remember that you should have vivado executable path into PATH variable because this action use vivado command.
.TP 
.B error  Vivado/2019.2/bin/loader: line 280: 27584 Killed  $RDI_PROG $@
This is a memory related problem that can be resolved by increasing the stack in VIVADO or, if it does not work, increasing the swap in your machine.
This was tested on a machine with 12GB RAM and initially 1GB swap (8GB at the end).
.RS 1.3i
.TP
.B Solution 1:
add another option to the VIVADO command: "-stack 2000" (i.e. run: vivado -stack 2000, also higher value) the vivado instruction to modify can be found in makefile of \fB pulpissimo/fpga/pulpissimo-BOARD \fR directory in the \fB all\fR recipe.
.TP
.B Solution 2:
Increase the swap in your machine to 8GB following these commands:
.RS 1.4i
 sudo swapoff -a
 sudo dd if=/dev/zero of=/swapfile bs=1G count=8
 sudo mkswap /swapfile
 sudo swapon /swapfile
 grep SwapTotal /proc/meminfo #to check if the swap is 8GB
.RE
.RE
.RE 

.TP
.B -d|--download
In order to download bitstream (because binary need flashing) into fpga we need the  JTAG usb connection. We also needs to set the board boot mode in \fB JTAG boot\fR usually using DIP switch on board as reported in the board User Guide ( see next board specific help).
This action use the makefile in pulpissimo/fpga to download bitstream.

.B Download common Error:
.RS 1.2i
.TP
.B Board is not detected
If the board is powerd on and connected, usually this is connected to Vivado driver, so execute \fB sudo ./install_drivers \fR in the directory 
.B {Xilinx_installation_dir}/Vivado/2019.2/data/xicom/cable_drivers/lin64/install_script
and reboot your machine
.RE
Now the fpga is read to emulate PULPissimo

.TP
.B -b|--build-sdk-openocd
This action create the 
.B BOARD.sh 
file in 
.B pulp-sdk/configs/fpgas/pulpissimo
directory and source it in order to configure the SDK for the FPGA platform, then 
.B configs/pulpissimo.sh
file is source to select correct microcontroller ( pulpissimo), finally the makefile in pulp-sdk directory is used to build the SDK. Also the json file
.B pulp-sdk/pulp-config/configs/fpgas/BOARD.json
is generated and used in building.
Finally all variable are saved in
.B .environ.env
file, these variable are setted to semplify some manual terminal operation, indeed once that -b action is done we could set this variable permanently using -e option that set variable in ~/.bash_profile that is sourced by ~/.bashrc at shell start. 

.TP
.B -c|--compile C_APPDIR
Compile test.c file founded in directory:
.B /pulp-rt-examples/\fBC_APPDIR\fR
creating the elf file:
.B /pulp-rt-examples/\fBC_APPDIR\fR/build/pulpissimo/test/test
That have the UART as default io peripheral.

.TP
.B -t|--terminals T_APPDIR
This Action can be performed if all previous Action are done, It open a series of terminals that connect board with openOCD to use gdb and UART to read data sent from board.
The option 
.B T_APPDIR
is used to select correct executable file located at
.B /pulp-rt-examples/\fBT_APPDIR\fR/build/pulpissimo/test/test.
The left-up terminal run openOCD using 
.B ./pulpissimo/fpga/pulpissimo-BOARD/CONNECTOR_NAME.cfg
configuration file, right-up terminal run gdb using 
.B $DIR/pulpissimo/fpga/pulpissimo-$BOARD/elf_run.gdb
script to run 
.B /pulp-rt-examples/$T_APPDIR/build/pulpissimo/test/test
executable. Instead other bottom terminal are USB interface, if the usb device is specified only one bottom terminal is displayed, if instead
.B all option is done all usb are screen in differen terminals.
The important thing to do is to create 
.B CONNECTOR_NAME.cfg 
file in 
.B ./pulpissimo/fpga/pulpissimo-$BOARD/
otherwhise openOCD give many errors!!!.

.TP 
.B -u|--usb-for-screen [ USB|all|i ]  (default USB=ttyUSB0)
Option to use with -t Action since define the usb used by board board to communicate with board. 
.B USB 
is for example ttyUSB0 (default value), if instead 
.B all 
is given, all ttyUSB from 0 to 5 are displayed ( supponing to have a USB 3.0 4-Port Hub ).
.B i 
option is used to show all usb device with their name.

.TP
.B -b|--board  BOARD (default BOARD=zcu102)
 This option is used to define the board to use, possible board are zcu102, zcu104, genesys2, zedboard, nexys or nexys_video. BOARD variable is  used in  -b,  -c  and  -t action. If you also use -r|--connector option, to set connector, be careful because BOARD option should be setted before -r option!!  otherwise default board will be used. For example:
.B pulp_app -o zcu104 -r olimex-arm-usb-tiny-h -b -c -t
is a valid command, while 
.B pulp\_app -r olimex-arm-usb-tiny-h -o zcu104 -b -c -t
don't work as you want since olimex-arm-usb-tiny-h.cfg file is searched in ./pulpissimo/fpga/pulpissimo-zcu102/ path since "zcu102" is the default BOARD.

.TP 
.B -r|--connector CONNECTOR_NAME (default CONNECTOR_NAME=olimex-arm-usb-tiny-h)
Set the base name of .cfg file used by openOCD to configure JTAG programmer and debugger used to program and debug the microcontroller. Indeed, JTAG is connected to FPGA JTAG, while PULPissimo JTAG  is mapped on another connector. It is important to observe that this connection can be made by using other different JTAG adapters, such as Digilent JTAG-HS2 or OLIMEX ARM-USB-OCD-H that are those ones suggested by the guide. Each adapter needs its own configuration file to setup the communication that will be saved in 
.B ./pulpissimo/fpga/pulpissimo-$BOARD/CONNECTOR_NAME.cfg.
In original repository there is only support for digilent-jtag-hs2 and olimex-arm-usb-ocd-h. If you have a different programmer/debugger you could use file in:
.B ./pulp-sdk/tools/riscv-openocd/tcl/interface/ftdi/
directory that contains files with configuration only for the openOCD programmer, instead file in ./pulpissimo/fpga/pulpissimo-$BOARD also contain fpga configuration. 
So you should merge this files, following this procedure:
.RS 1.2i
.TP 
.B 1
Copying file ./pulp-sdk/tools/riscv-openocd/tcl/interface/ftdi/CONNECTOR_NAME.cfg into  ./pulpissimo/fpga/pulpissimo-$BOARD/
.TP
.B 2
Copying text from  ./pulpissimo/fpga/pulpissimo-$BOARD/digilent-jtag-hs2.cfg file from the line: 
.B set _CHIPNAME riscv 
to the end and appending it in your new config file (in ./pulpissimo/fpga/pulpissimo-$BOARD/) . 
.RE

.SH BOARDS
.TP 
.B zcu102
.RS 1.2i
.TP
.B USB/JTAG/UART micro-USB connector (J2)
 This is used to download the bitstream into fpga.
.TP
.B USB/UART micro-USB connector (J83)
This port is used to UART connection of PULPissimo with the computer.
.TP 
.B Olimex_OpenOCD_JTAG_ARM-USB-TINY-H programmer and debugger (J55)
This device is used to load the application into core's memory and debug PULPissimo's core that has its own JTAG interface. Indeed, J83 port is connected to FPGA JTAG, while PULPissimo JTAG  is mapped on J55 connector. It is important to observe that this connection can be made by using other different JTAG adapters, such as Digilent JTAG-HS2 or OLIMEX ARM-USB-OCD-H that are those ones suggested by the guide. Each adapter needs its own configuration file to setup the communication. Following PULPissimo guide , you should configure the correct connection between JTAG_ARM-USB-TINY-H and PULPissimo JTAG.
.RE
.PP
.SH SEE ALSO
.BR pulp_install 1 
.BR pulp_script 1
.SH BUGS            
.SH CREDITS 
Created by Elia Ribaldone, Luca Fiore and Marcello Neri


pulp_script/pulp_app.sh
#!/bin/bash 

#############################################################################
####### FUNCTIONS ###########################################################
setJsonBoard(){
	# This function create the correct json file
	# in pulp-sdk/pulp-config/configs/fpgas/ directory
	# used to build sdk
	# The second file created is the board file used to
	# set board variable and platform to use setted as fpga
	BOARD=$1
	cd ./pulp-config/configs/fpgas/
        mv genesys2.json $BOARD.json
        sed -i 's/genesys2/$BOARD/g' $BOARD.json
	cd ../../../
	createBoardFile $BOARD
	mv $BOARD.sh ./configs/fpgas/pulpissimo/
}

usbInfo(){
	# This function is used to give information about usb name
	for sysdevpath in $(find /sys/bus/usb/devices/usb*/ -name dev); do
	    (
		syspath="${sysdevpath%/dev}"
		devname="$(udevadm info -q name -p $syspath)"
		[[ "$devname" == "bus/"* ]] && exit
		eval "$(udevadm info -q property --export -p $syspath)"
		[[ -z "$ID_SERIAL" ]] && exit
		echo "/dev/$devname - $ID_SERIAL"
	    )
	done
}

UsageExit()
{
	# Print help and exit
	echo -e "\
pulp_app [OPTION]   Is a bash script able to create pulp bitstream,
		    upload it into board compile application and run
		    it into board, actually zcu102 is the  default
		    board and hello is the default application.
OPTION
	-s|--bitstream
		create bitstream and binary file for flashing, this
		option use vivado
	-d|--download
		download bitstream into board using
		make -C pulpissimo-BOARD download
	-b|--build-sdk-openocd
		build the sdk and opeocd patch
	-c|--compile C_APPDIR
		create cross compiled test elf file from test.c 
		file founded in ./pulp-rt-examples/\$C_APPDIR
	-t|--terminal T_APPDIR
		if the  terminal with openOCD, gdb and screen, 
		for screen is possible to select usb number using
		-u option and -r for connector option. The file
		tested is pulp-rt-examples/\$T_APPDIR/build/pulpissimo/test/test
		created by compilation using -c option.
	-e|--export-variables
		After -b option the environment varible  are setted in .environ.env 
		file, so this action set variable find in .environ.env in ~/.bash_profile
		, in this way each time that a shell is open the variable are setted.
		After the execution of the action you should restart shell 
		to have variable setted. 
	-h|--help
		print this help
	-u|--usb-for-screen USB|all|i
		selection of usb for screen connection example:
		-u ttyUSB0
		all option istead screen all usb
		i option show corrently usb and their name
	-o|--board BOARD
		selection of target board, possible board are
		zcu102, zcu104, genesys2, zedboard, nexys or nexys_video
	-r|--connector CONNECTOR_NAME
		CONNECTOR_NAME should be the name of jtag programmer and debugger
		used to upload and debug application on pulpissimo. This 
		connector should support openOCD. This name is used to call 
		./pulpissimo/fpga/pulpissimo-\$BOARD/\$CONNECTOR_NAME.cfg 
		as configuration of openOCD.
	
"
	exit 1;	
}

createBoardFile(){
	# This board file is used to set board
	BOARD=$1
	echo "#!/bin/bash
BOARD=VARIABLE

scriptDir=\"\$\(dirname \"\$\(readlink -f \"\$\{BASH_SOURCE[0]}\")\")\"

export PULP_CURRENT_CONFIG=pulpissimo@config_file=fpgas/pulpissimo/$BOARD.json

export PULP_CURRENT_CONFIG_ARGS=platform=fpga

if [ -e \${scriptDir}/../../../init.sh ]; then
    source \${scriptDir}/../../../init.sh
fi
" > $1.sh
}
 
# after installation BASHLIB option will be exported by 
# .bash_profile and used here to find ccommon.sh file
# that is placed in $BASHLIB directory by installer
source $BASHLIB/ccommon.sh

# Regular Colors
Black='\033[0;30m'        # Black
Red='\033[0;31m'          # Red
Green='\033[0;32m'        # Green
Yellow='\033[0;33m'       # Yellow
Blue='\033[0;34m'         # Blue
Purple='\033[0;35m'       # Purple
Cyan='\033[0;36m'         # Cyan
White='\033[0;37m'        # White

DIR=`pwd`
#sudo chmod 2775 /usr/bin/screen
#sudo pkill -9 screen



### VARIABLES
BOARD="zcu102"
BITSTREAM=0
DOWNLOAD=0
COMPILE=0
TERMINALS=0
NOTHING=1
USB="ttyUSB0"
BUILD_SDK_OPENOCD=0
EXPORT=0
CONNECTOR_NAME="olimex-arm-usb-tiny-h"

exitflag=0
LOG=$DIR"/pulp_app_log"
ENVIRON_FILE=$DIR/.environ.env
###

#######################################################################
####### COMMAND LINE ARGUMENT HANDLE ##################################

TEMP=`getopt -o sdc:t:ehbu:o: --long bitstream,download,compile:,terminals:,--export-variables,help,build-sdk-openocd,usb-for-screen:,board: -- "$@"`

eval set -- "$TEMP"

while true; do
	case $1 in
		###### ACTION ##########
		-s|--bitstream)
			BITSTREAM=1	
			NOTHING=0
			shift
			;;
		-d|--download)
			DOWNLOAD=1	
			NOTHING=0
			shift
			;;
		-b|--build-sdk-openocd)
			BUILD_SDK_OPENOCD=1
			NOTHING=0
			shift
			;;
	       	-c|--compile)
			COMPILE=1
			shift
			C_APPDIR=$1
			echo $1
			NOTHING=0
			shift
		-t|--terminals)
			TERMINALS=1
			shift
			T_APPDIR=$1
			echo $1
			NOTHING=0
			shift                                                                           
			;;
		-e|--export-variables)
			NOTHING=0
			EXPORT=1
			shift
			;;
		-h|--help)
			UsageExit
			;;
		######## OPTION #######################à
		-u|--usb-for-screen)
			shift
			USB=$1
			echo $1
			shift
			;;
		-o|--board)
			shift
                        case $1 in
                                zcu102)
                                        ;;
                                zcu104)
                     USB          ;;
                                genesys2)
                                        ;;
                                zedboard)
                                        ;;
                                nexys)
                                        ;;
                                nexys_video)
                                        ;;
                                default)
                                        echo "Error board: zcu102, zcu104, genesys2, zedboard, nexys or nexys_video"
                                        exit 1;
                                        ;;
                        esac
                        BOARD=$1
			shift
			;;
		-r|--connector)
			shift
			cd ./pulpissimo/fpga/pulpissimo-$BOARD
			if test -f $1.cfg; then
				CONNECTOR_NAME=$1
			else
				Print "r" "At ./pulpissimo/fpga/pulpissimo-$BOARD there isn't openocd configuration file $1.cfg!! create it"
				UsageExit
			fi
			;;
		--)
			break;;
		*)
			UsageExit
			;;
	esac		
done

if [[ $NOTHING -eq 1 ]] && [[ $USB != "i" ]]; then
	echo $Red"Please give at least an argument!!"
	UsageExit
fi

#######################################################################
###### PROGRAM ######################################################

# Setting environment variable
# variable previous setted by -b option in $ENVIRON_FILE will be saved in ~/.bash_profile file
# in order to be available in shell after restart.
if [[ $EXPORT  -eq 1 ]]; then
	if test -f "$ENVIRON_FILE"; then
		Print "e" "Set environment var"
		while read var; do
			export_var "$(echo $var| cut -d "=" -f 1)" "$(echo $var| cut -d "=" -f 2)";
			echo "$var exported";
		done < $ENVIRON_FILE
		Print "e" "Restart shell to use variable!!"
	else
		echo "Error environment not setted!!!"
		exitflag=1
	fi 
fi

# Creation of log directory for mon_run output
mkdir -p $LOG

if [[ $BITSTREAM -eq 1 ]]; then
	Print "b" " Bitstream generation:"
	cd $DIR/pulpissimo
	##### Generate the necessary synthesis include scripts #####
	#It will parse the ips_list.yml using the PULP IPApproX IP
        # management tool to generate tcl scripts for all the IPs used in 
	# the PULPissimo project. These files are later on sourced by Vivado
        #to generate the bitstream for PULPissimo.
	./update-ips
	# generate bitstream 
	cd fpga
	mon_run "make clean_$BOARD" "$LOG/bitstream_log.txt" 1 $LINENO
	
	mon_run "make $BOARD" "$LOG/bitstream_log.txt" 0 $LINENO
	cd $DIR
fi

if [[ $DOWNLOAD -eq 1 ]]; then
	# download of bistream into fpga, remember that this isn't a flashing, 
        # if you restart fpga all configuration will be lost and -d action
	# should be redone.	

	# waiting for user to connect cable for programming
	Action "Have you connected usb cables for riscv download at port J2 of $BOARD?"
	cd $DIR/pulpissimo/fpga
	Print "f" " Flashing fpga"
	mon_run "make -C pulpissimo-$BOARD download" "$LOG/flashing_log.txt" 1 $LINENO
fi

if [[ $BUILD_SDK_OPENOCD -eq 1 ]] || [[ $COMPILE -eq 1 ]] ; then
	# Setting of env varible 
	cd $DIR/pulp-sdk
	setJsonBoard $BOARD
	source configs/pulpissimo.sh
	source configs/fpgas/pulpissimo/$BOARD.sh
        mon_run "sudo apt-get install autoconf automake texinfo make libtool pkg-config libusb-1.0 libftdi1 libusb-0.1" "$LOG/build_sdk_openocd_dep" 1 $LINENO
	source sourceme.sh
	export PULP_RISCV_GCC_TOOLCHAIN="/opt/riscv"
	export OPENOCD="$DIR/pulp-sdk/pkg/openocd/1.0/bin"
fi

if [[ $BUILD_SDK_OPENOCD -eq 1 ]]; then
	# Building of sdk and openocd patch
	mon_run "./pulp-tools/bin/plpbuild checkout build --p openocd --stdout" "$LOG/checkout.txt" 1 $LINENO
	mon_run "make all" "$LOG/make_all.txt" 1 $LINENO
	env | grep -e "PULP\|OPENOCD" > $ENVIRON_FILE
fi

if [[ $COMPILE -eq 1 ]]; then
	# compiling application
	Print "c" " Compiling"
	cd $DIR/pulp-rt-examples/$C_APPDIR
	echo $(pwd)
        mon_run "make clean all" "$LOG/make_log.txt" 1 $LINENO
fi

if [[ $USB == "i" ]] ; then
	usbInfo
fi
if [[ $TERMINALS -eq 1 ]] && [[ $exitflag -eq 0 ]]; then
	Action "Have you connected usb cables for UART  at port J38 of zcu102 and ARM-USB_TINY-H connector to J55 port?"
  		
  	cd $DIR
        # openOCD terminal
	gnome-terminal --geometry=70x15+0+0 -- bash -c "$OPENOCD/openocd -f $DIR/pulpissimo/fpga/pulpissimo-$BOARD/$CONNECTOR_NAME.cfg; exec bash"
	
	# gdb terminal
	gnome-terminal --geometry=70x15+1000+0 -- bash -c "$PULP_RISCV_GCC_TOOLCHAIN/bin/riscv32-unknown-elf-gdb -x $DIR/pulpissimo/fpga/pulpissimo-$BOARD/elf_run.gdb ./pulp-rt-examples/$T_APPDIR/build/pulpissimo/test/test; exec bash"
	
	# Various UART terminal
	if [[ $USB != "all" ]]; then
		sudo chmod 777 /dev/$USB
		sudo gnome-terminal --geometry=70x15+1000+355 -- bash -c "sudo screen /dev/$USB 115200; exec bash"
	else

		for i in {0...5}; do
			sudo chmod 777 /dev/ttyUSB$i
		done
		sudo gnome-terminal --geometry=70x15+0+355 -- bash -c "sudo screen /dev/ttyUSB0 115200; exec bash"
		sudo gnome-terminal --geometry=70x15+200+355 -- bash -c "sudo screen /dev/ttyUSB1 115200; exec bash"
		sudo gnome-terminal --geometry=70x15+300+355 -- bash -c "sudo screen /dev/ttyUSB2 115200; exec bash"
		sudo gnome-terminal --geometry=70x15+400+355 -- bash -c "sudo screen /dev/ttyUSB3 115200; exec bash"
		sudo gnome-terminal --geometry=70x15+500+355 -- bash -c "sudo screen /dev/ttyUSB4 115200; exec bash"
		sudo gnome-terminal --geometry=70x15+600+355 -- bash -c "sudo screen /dev/ttyUSB5 115200; exec bash"
	fi

fi



